packages/nextjs/src/config/webpack.ts

```diff
@@ -98,26 +98,34 @@ export function constructWebpackConfigFunction(
       ],
     });
 
-    let pagesDirPath: string;
+    let pagesDirPath: string | undefined;
     const maybePagesDirPath = path.join(projectDir, 'pages');
+    const maybeSrcPagesDirPath = path.join(projectDir, 'src', 'pages');
     if (fs.existsSync(maybePagesDirPath) && fs.lstatSync(maybePagesDirPath).isDirectory()) {
-      pagesDirPath = path.join(projectDir, 'pages');
-    } else {
-      pagesDirPath = path.join(projectDir, 'src', 'pages');
+      pagesDirPath = maybePagesDirPath;
+    } else if (fs.existsSync(maybeSrcPagesDirPath) && fs.lstatSync(maybeSrcPagesDirPath).isDirectory()) {
+      pagesDirPath = maybeSrcPagesDirPath;
     }
 
-    let appDirPath: string;
+    let appDirPath: string | undefined;
     const maybeAppDirPath = path.join(projectDir, 'app');
+    const maybeSrcAppDirPath = path.join(projectDir, 'app');
     if (fs.existsSync(maybeAppDirPath) && fs.lstatSync(maybeAppDirPath).isDirectory()) {
-      appDirPath = path.join(projectDir, 'app');
-    } else {
-      appDirPath = path.join(projectDir, 'src', 'app');
+      appDirPath = maybeAppDirPath;
+    } else if (fs.existsSync(maybeSrcAppDirPath) && fs.lstatSync(maybeSrcAppDirPath).isDirectory()) {
+      appDirPath = maybeSrcAppDirPath;
     }
 
-    const apiRoutesPath = path.join(pagesDirPath, 'api');
+    const apiRoutesPath = pagesDirPath ? path.join(pagesDirPath, 'api') : undefined;
+
+    const middlewareLocationFolder = pagesDirPath
+      ? path.join(pagesDirPath, '..')
+      : appDirPath
+      ? path.join(appDirPath, '..')
+      : projectDir;
 
-    const middlewareJsPath = path.join(pagesDirPath, '..', 'middleware.js');
-    const middlewareTsPath = path.join(pagesDirPath, '..', 'middleware.ts');
+    const middlewareJsPath = path.join(middlewareLocationFolder, 'middleware.js');
+    const middlewareTsPath = path.join(middlewareLocationFolder, 'middleware.ts');
```

Comment:
Note that some people use the convention of using `.tsx` for all typescript files, even if they don't contain JSX
Next.js allows middlewares with `.tsx` extension (and probably also jsx, not tested)

Perhaps it would be a good idea to accept middlewares with any of the `pageExtensions`

Reply:
Ugh, you might be right. Once again not in line with what the Next.js docs say: https://nextjs.org/docs/app/building-your-application/routing/middleware#convention